Support dedicated filter text in "completions" options (#4418)
mawww/kakoune: doc/pages/options.asciidoc

    `set -remove` removes the given specs from the list. +

*completions*::
    a list of `<text>[|<filter text>]|<select cmd>|<menu text>` candidates,
    except for the first element which follows the
    `<line>.<column>[+<length>]@<timestamp>` format to define where the
    completion apply in the buffer.
    Any `|` or `\` characters that occur within the `<text>`,
    `<filter text>`, `<select cmd>`, or `<menu text>` fields should be
    escaped as `\|` or `\\`.

    Kakoune will filter away candidates that do not match what the user types.

>     Candidates are matched against the optional `<filter text>`, or `<text>`.

Author: yep, I added a commit that tries to explain this

> - the completion menu displays "filter text" (if provided) and/or "text"

Currently that's an "or", but at least for the Julia completion of Unicode symbols it would definitely work well to display `\xrat (℞)` here.

I did not yet find other examples where the text-to-insert is different from the filter text.
If the filter text can be very similar to the text-to-insert, it may not be wise to display both.

>     Candidates are matched against the optional `<filter text>`, or `<text>`.

Reviewer: The more general form of completions is "type the name of something, get something else". It could be something shorter (like typing "Smiling Face with Open Mouth and Tightly-Closed Eyes" to get 😆), or it could be something much longer (like typing "defprop" to get a snippet that defines a property access method — I know @alexherbo2 has been working on something like this).

If I had to pick a convention for people to follow, it would probably be that the completion menu itself displays the text that I need to type to narrow down the list of completions, while the menu text (if short) or `info -menu` (if long) is used to preview the thing that will be inserted when I accept an option.

>     Candidates are matched against the optional `<filter text>`, or `<text>`.

Author: Right, clients that use filter-text can use the menu-text and the on-select handler to display what would be inserted.

    `set -remove` removes the given specs from the list. +

*completions*::
    a list of `<text>[|<filter text>]|<select cmd>|<menu text>` candidates,
    except for the first element which follows the
    `<line>.<column>[+<length>]@<timestamp>` format to define where the
    completion apply in the buffer.
    Any `|` or `\` characters that occur within the `<text>`,
    `<filter text>`, `<select cmd>`, or `<menu text>` fields should be
    escaped as `\|` or `\\`.

    Options of this type are are meant to be added to the `completers`
    option to provide insert mode completion.  Candidates are shown if the

>     text typed by the user (between `<line>.<colum>` and the cursor) is a

Reviewer: ```suggestion
    text typed by the user (between `<line>.<column>` and the cursor) is a
```

>     prefix of `<filter text>` (if provided) or `<text>`.

Reviewer: I'm not sure there's a concise word that describes Kakoune's fuzzy matching, but "substring" might be better than "prefix".

Author: I changed it to [subsequence](https://en.wikipedia.org/wiki/Subsequence) which also matches the term used by the implementation (see `subsequence_match_smart_case`)

    For each remaining candidate, the completion menu displays `<filter text>`
    (if provided) or `<text>`, followed by `<menu text>`, which can contain
    Markup (see <<faces#markup-strings,`:doc faces markup-strings`>>).

>     If the user selects an item from the completion menu, the text they typed

Reviewer: ```suggestion
    As the user selects items from the completion menu, the text they typed
```

With the original wording, it sounded like these actions occurred once the user had accepted a completion, rather than just while they cycle through the completione.
